src/util/util.ts

```diff
@@ -349,29 +364,105 @@ export function runWithListeningChildProcesses<In>(
                     }
                     child.kill();
                 } else {
-                    child.send(inputs[inputIndex]);
+                    currentInput = inputs[inputIndex];
                     inputIndex++;
+
+                    if (oldCrashRecoveryState === CrashRecoveryState.RetryWithMoreMemory) {
+                        // succeeded when running with more memory. Restart the child with less memory
+                        restartChild();
+                    } else {
+                        child.send(currentInput);
+                    }
                 }
-            });
-            child.on("disconnect", () => {
-                if (inputIndex !== inputs.length) {
-                    fail();
+            };
+
+            const onClose = (code: number, signal: string) => {
+                if (rejected || inputIndex === inputs.length) {
+                    return;
                 }
-            });
-            child.on("close", () => { assert(rejected || inputIndex === inputs.length); });
-            child.on("error", fail);
+
+                if (crashRecovery && code === 134 && signal === null) {
+                    switch (crashRecoveryState) {
+                        case CrashRecoveryState.Normal:
+                            crashRecoveryState = CrashRecoveryState.Retry;
+                            if (handleCrash) {
+                                handleCrash(currentInput, crashRecoveryState);
+                            }
+                            restartChild();
+                            return;
+                        case CrashRecoveryState.Retry:
+                            // skip crash recovery if we're already passing a value for --max_old_space_size that is >= crashRecoveryMaxOldSpaceSize
+                            if (maxOldSpaceSize < crashRecoveryMaxOldSpaceSize) {

```

Comment:
why the secondary check? is 4096 just a hard limit right now, even if the command line specifies more?

Reply:
or is there a second retry if the command line specifies <4096?

Reply:
If we specify `--max_old_space_size` to the root process then it wouldn't make sense to retry with "more" memory, since `fork` copies the `execArgv` of the parent to each child (i.e. each child would already be running with `--max_old_space_size` set).